Add introduction to the section on variants and inductive types in the reference manual

herbelin opened the pull request:
This is a preliminary proposal to add an introductive header to the Variant and Inductive sections of the reference manual.

It is probably not ok as such. For instance, it should probably be decided first what comes first in the explanations between `Variant` and `Inductive`, where is the syntax given, or whether it is duplicated, or if the sections have to be both readable independently, etc. So, waiting for an opinion of the reference manual coordinators.

Fixes / closes #17786

- [x] Added / updated **documentation**.

proux01 commented:
I like it. Currently `Variant`comes before `Inductive` which I like as it enables to introduce `match` before `fix` but that's just my personnal opinion.

proux01 commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -24,6 +28,33 @@ Variants
+     Variant reflect (P : Prop) : bool -> Set :=
+     | ReflectT : P -> reflect P true
+     | ReflectF : ~ P -> reflect P false.
+     Variant eq (A : Type) (x : A) : A -> Prop :=  eq_refl : eq x x.
```
Do we already know about implicit argument at this point? Otherwise maybe
```suggestion
     Variant eq (A : Type) (x : A) : A -> Prop :=  eq_refl : @eq A x x.
```
could help?
Would those two type require a bit more explanation?

SkySkimmer replied:
That change still needs knowledge of implicit arguments to explan the @ 
Maybe disable automatic implicits instead

proux01 commented:
BTW, does this depend on another PR? it seems to include quite a bit of unrelated material.

herbelin commented:
> BTW, does this depend on another PR? it seems to include quite a bit of unrelated material.

The unrelated material was a mistake. Now removed.

Implicit arguments of `eq`: added.

> Would those two type [eq and reflect] require a bit more explanation? 

If eventually kept, they could indeed be accompanied with more explanations.

jfehrle commented on doc/sphinx/language/core/inductive.rst:
```diff
@@ -3,6 +3,10 @@ Inductive types and recursive functions
 
 .. _gallina-inductive-definitions:
 
+The :cmd:`Inductive` command allows to define types by case on the inhabitants of the type. On the contrary of types defined by the command :cmd:`Variant`, constructors can recursively have arguments in the type being defined. So, inductive types include natural numbers, lists, well-founded trees, etc. Inhabitants of inductive type can recursively nest only a finite number of constructors. So, th ...
```
```suggestion
The :cmd:`Inductive` command allows defining types by cases on the inhabitants of the type. In contrast, types defined by the :cmd:`Variant` command, constructors can recursively have arguments in the type being defined. Inductive types include natural numbers, lists and well-founded trees. Inhabitants of inductive type can recursively nest only a finite number of constructors. So, they are well-founded. This distinguishes them from :cmd:`CoInductive` types, whose constructors can be infinitely nested. In Coq, :cmd:`Variant` types thus correspond to the common subset of inductive and coinductive types that are non-recursive.
```

Please split this into multiple lines that are ~80 characters long.

This needs some work.
"define types by case on the inhabitants of the type" doesn't make sense to me unless you mean "inhabitants of another type"
How about defining "well-founded" and adding maybe add it to the glossary?
Also, would you give a concrete example of a coinductive that's infinitely nested?  That's not at all obvious.

IIRC there is not a clear explanation (with example) of the difference between a coinductive and an inductive.  Could you add that?

jfehrle commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,6 +6,10 @@ Variants and the `match` construct
 Variants
 --------
 
+The :cmd:`Variant` command allows to define types by case on the inhabitants of the type. This includes enumerated types from programming languages, such as Booleans, characters or even the degenerate cases of the unit and empty types. Variant types more generally include enumerated types with arguments or even enumerated types with parametric arguments such as option types, sum types, etc. It al ...
```
```suggestion
The :cmd:`Variant` command allows defining types by case on the inhabitants of the type. This includes enumerated types from programming languages, such as Booleans, characters or even the degenerate cases of the unit and empty types. Variant types more generally include enumerated types with arguments or even enumerated types with parametric arguments such as option types, sum types, etc. It also includes predicates or type families defined by cases such as the Boolean reflection or equality predicates. Observing the form of the inhabitant of a variant type is done by case analysis using the :cmd:`match` expression. Each case is described by what is called a :g:`constructor` of the type. For instance, Booleans have two constructors: :g:`true` and :g:`false`.
```

Variant types are not recursive (your examples boolean and character are not)?
Maybe add "inhabitant" to the glossary if it's essential to use that word here.
"Constructor" should be in the glossary and defined more explicitly and less tersely.

Please split into multiple lines.

jfehrle reviewed (commented):
Needs some work.  I'll have more time to look at it when I'm back from vacation on July 11.

Zimmi48 commented:
Thanks a lot! :sparkles: 

> For instance, it should probably be decided first what comes first in the explanations between `Variant` and `Inductive`, where is the syntax given, or whether it is duplicated, or if the sections have to be both readable independently, etc. So, waiting for an opinion of the reference manual coordinators.

Like Pierre said, it was a choice of the restructuring of the refman to put variants first to explain the `match` construct then. It makes sense that the two chapters cross-reference each other (like you did) and that a reader of the chapter on inductive or co-inductive types will have to refer to the chapter on variants to learn about the `match` construct.

The syntax has some duplication, of course, but it should stay like it is today (easier to read).

>Needs some work.

This can definitely be improved further, and it is great if the interactions between the fresh eye of Jim and the expert eye of Hugo can lead to an even better and crystal clear introduction, but let's not forget that perfect is the enemy of good, so let's be careful to not let the PR stall and merge it then as it is already a much-needed improvement over what we have today.

herbelin commented:
@jfehrle: Thanks for the feedback.

> Maybe add "inhabitant" to the glossary

One problem in logic is that some words, such as "element", have a formal meaning in set theory. So, we tend to use alternative terms such as value, member, inhabitant. Among those, is there one which looks easier to intuitively grasp? 

In any case, the chosen terminology can be added to the section introducing [types](https://coq.inria.fr/refman/language/core/basic.html#term-type).

> "Constructor" 

If not done, this has to be in the glossary. And with a clear definition.

>  Also, would you give a concrete example of a coinductive that's infinitely nested? That's not at all obvious.

If not already done, it can be done. Short example: `CoInductive nat := S : nat -> nat.` which include `S (S (S (S ...))))`.

My primary questions are about the organisation of the sections:
- Are the sections supposed to be readable in any order, or should the `Inductive` section be read first, or the `Variant` section first? E.g. Pierre liked to have Variant be read first. This has an impact e.g. on where to put the grammar.
- Should the header be for Variant alone or for match also, or should there be two headers? Is there already a convention regarding it? And same question for the Inductive section?

herbelin commented:
Small update: I split the lines, linked to the definition of Leibniz equality in "writing-proofs/equality.rst", added a smoother transition with the previous first lines of the variants section, replaced "witness" by "inhabitants" in basic. rst (as a consequence, this *defines* inhabitants, which I observed was already used at other places of the refman), added the coinductive example of streams.
Probably more smoothing to do.

Zimmi48 commented:
> One problem in logic is that some words, such as "element", have a formal meaning in set theory. So, we tend to use alternative terms such as value, member, inhabitant. Among those, is there one which looks easier to intuitively grasp?

I personally prefer the use of "inhabitant" because I believe it is more standard.

Zimmi48 commented:
>Are the sections supposed to be readable in any order, or should the Inductive section be read first, or the Variant section first? E.g. Pierre liked to have Variant be read first. This has an impact e.g. on where to put the grammar.

They can be read in any order, but the natural order is to read Variant first. In any case, if the reader chooses to start by Inductive, there should be sufficient links to Variant so that these links can be followed when needed.

>Should the header be for Variant alone or for match also, or should there be two headers? Is there already a convention regarding it? And same question for the Inductive section?

The main header should be for the entire content of the chapter, this is why it is currently "Variants and the match construct", which is the split in two sections, one on variants, and one of the match construct.

Zimmi48 commented:
>as a consequence, this defines inhabitants, which I observed was already used at other places of the refman

Don't hesitate to use the `:term:` syntax to link from these occurrences to their definition.

jfehrle commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -24,6 +43,37 @@ Variants
+      Variant option (A : Type) : Type := None : option A | Some : A -> option A.
+      Variant sum (A B : Type) : Type := inl : A -> sum A B | inr : B -> sum A B.
+
+  *Boolean reflection* is a relation reflecting under the form of a
```
FWIW, "reflection" is probably worth adding to the glossary at some point.

herbelin replied:
It is somehow more specfic to the library than to the logic.

jfehrle commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -24,6 +43,37 @@ Variants
+     | ReflectT : P -> reflect P true
+     | ReflectF : ~ P -> reflect P false.
+
+  :term:`Leibniz Equality` is another example of variant type.
```
Actually, it's defined as a Inductive: `Inductive eq (A:Type) (x:A) : A -> Prop :=
    eq_refl : x = x :>A`

Zimmi48 replied:
The actual keyword that is used `Inductive` or `Variant` does not matter. This is a variant because it is not recursive. Maybe we should make this clearer in the doc...

herbelin replied:
We are facing a dilemma here. It is indeed artificial to use `Inductive` for a type which is neither inductive nor coinductive but `Variant` does not define `eq_ind`/`eq_rect`, which are pervasively used.

So, we would need to add a "Set Nonrecursive Elimination Schemes" beforehand. But why not, if it helps to make thinks clearer.

jfehrle replied:
Probably worth mentioning the use of Inductive to create XX_ind and XX_rect as an alternative in the description of Variant.

jfehrle commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
+defined, the type becomes recursive, in which case it can be either
+:cmd:`Inductive` or :cmd:`CoInductive`. The terminology :cmd:`Variant`
+is thus specifically used to indicate a non-recursive type defined by
+constructors (for instance, natural numbers, lists or streams cannot
```
"stream" (an infinite sequence of numbers?) may not be familiar to your average software engineer.  I didn't know it.

herbelin replied:
In Coq, streams are the following type:
```coq
CoInductive Stream (A : Type) := { hd : A ; tl : Stream A }.
```
This indeed generalizes infinite sequence of numbers to infinite sequence of any kind of objects.

Then, [coinductive types](https://en.wikipedia.org/wiki/Coinduction#Codata) is a generalization of streams.

jfehrle reviewed (commented):
A few further thoughts

It's fine to use "inhabitant" provided we define it.

There's still room for further refinement, but I would need time to read more of what's already in the doc in order to make sensible suggestions.

jfehrle commented:
> The main header should be for the entire content of the chapter

Really should be "introduction" rather then "header".  A definition of header:

> A header is a separate bit of text at the top of a printed page. A header might be the title of the novel you're reading, which is repeated on each page of the book. One kind of header — also called a "page header" — is a chapter title, which is often printed at the top of every page.

jfehrle commented:
To fix the refman build, add `abort` on this coqtop line in variants.rst:

```
.. coqtop:: in

   Inductive bool : Type := true : bool | false : bool.
```

```
Extension error:
/builds/coq/coq/_build/default/doc/sphinx/language/core/variants.rst:189: Error while sending the following to coqtop:
    Inductive bool : Type := true : bool | false : bool.
  coqtop output:
    Toplevel input, characters 0-52:
    > Inductive bool : Type := true : bool | false : bool.
    > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    Error: bool already exists.

Zimmi48 commented:
> To fix the refman build, add `abort` on this coqtop line in variants.rst:

I don't think `abort` would solve this. It should probably use `reset`.

herbelin commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
+
+When constructors recursively take arguments in the type being
+defined, the type becomes recursive, in which case it can be either
+:cmd:`Inductive` or :cmd:`CoInductive`. The terminology :cmd:`Variant`
+is thus specifically used to indicate a non-recursive type defined by
+constructors (for instance, natural numbers, lists or streams cannot
+be defined using :cmd:`Variant`).
```
I would prefer:
```
:cmd:`Inductive` or :cmd:`CoInductive`. The keyword :cmd:`Variant`
is reserved for non-recursive types.  Natural numbers, lists and streams cannot
be defined using :cmd:`Variant`.
```
if ok for you.

herbelin commented:
New version with some uses of ``term:`...` ``, and with a way to prevent the "bool already exists" error.

Zimmi48 commented on doc/sphinx/language/core/inductive.rst:
```diff
@@ -3,6 +3,22 @@ Inductive types and recursive functions
 
 .. _gallina-inductive-definitions:
```
This label should stay right before the corresponding header. This is what makes the build currently fail.

Zimmi48 commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,29 @@ Variants and the `match` construct
+is reserved for non-recursive types. Natural numbers, lists or streams cannot
+be defined using :cmd:`Variant`.
+
+.. cmd:: Variant @ident {? @cumul_univ_decl } {* @binder } {? %| {* @binder } } {? : @type } := {? %| } {+| @constructor } {? @decl_notations }
```
There are 15 uses of `ident_decl` in the refman. If you think that it creates confusion and should be unfolded, we should do it for all the occurrences (at the risk of making grammar description longer). You can test this change by adding `ident_decl` to the SPLICE list of `common.edit_mlg` (currently starting on line 2260) and running `make doc_gram_rsts`. This change which doesn't match what is in `common.edit_mlg` / `orderedGrammar` is the reason for the linter failure.

herbelin replied:
I was needing a way to refer to the name "ident" of the variant type. I followed the example of the Inductive section where ident_decl was already split. I have to particular preference otherwise.

Zimmi48 replied:
It is split in `Inductive` because it uses `cumul_univ_decl` which is not used anywhere else. Typically the way we refer to production names that do not appear directly is to write "the `ident` (in `ident_decl`)".

herbelin replied:
I see. Done. Hopefully as expected. If not, I can still change.

herbelin commented:
I had replaced `witness` by `inhabitant` in `basic.rst` and in the index but the compilation is now failing on witness not any more in the index.

Any objection to let `witness` out of the index? The first occurrence (in `sorts.rst`) could be kept and understood infornally (w/o reference to the index), while the second occurrence (in `evars.rst`) could nicely be replaced with `inhabitant` (together with a link to the index)??

If ok, I'll push this change.

Zimmi48 commented on doc/sphinx/language/core/inductive.rst:
```diff
@@ -1,6 +1,22 @@
 =======================================
 
+The :cmd:`Inductive` command allows to define types by case on the
+inhabitants of the type. On the contrary of types defined by the
```
Should we link to the term definition here as well?
```suggestion
:term:`inhabitants <inhabitant>` of the type. On the contrary of types defined by the
```

Zimmi48 commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
 --------
 
+The :cmd:`Variant` command allows defining types by listing
+the :term:`inhabitants` of the type. This includes enumerated types from
```
This is the reason why the compilation currently fails:

```suggestion
the :term:`inhabitants <inhabitant>` of the type. This includes enumerated types from
```

Zimmi48 commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
+types with parametric arguments such as option types, sum types,
+etc. It also includes predicates or type families defined by cases
+such as the Boolean reflection or equality predicates. Observing the
+form of the term:`inhabitants` of a variant type is done by case analysis
```
```suggestion
form of the :term:`inhabitants <inhabitant>` of a variant type is done by case analysis
```

jfehrle commented on doc/sphinx/language/core/inductive.rst:
```diff
@@ -1,6 +1,22 @@
 Inductive types and recursive functions
 =======================================
 
+The :cmd:`Inductive` command allows to define types by case on the
+:term:`inhabitants <inhabitant>` of the type. On the contrary of types defined by the
+command :cmd:`Variant`, constructors can recursively have arguments in
+the type being defined. So, inductive types include natural numbers,
+lists, well-founded trees, etc. Inhabitants of inductive type can
+recursively nest only a finite number of constructors. So, they are
+well-founded. This distinguishes them from :cmd:`CoInductive` types,
+such as streams, whose constructors can be nested infinitely. In Coq,
+:cmd:`Variant` types thus correspond to the common subset of inductive
+and coinductive types that are non-recursive.
```
```suggestion
The :cmd:`Inductive` command allows defining types by cases on the
:term:`inhabitants <inhabitant>` of the type. In contrast, in types defined by the
:cmd:`Variant` command, constructors can recursively have arguments in
the type being defined. Inductive types include natural numbers,
lists and well-founded trees. Inhabitants of inductive types can
recursively nest only a finite number of constructors. So, they are
well-founded. This distinguishes them from :cmd:`CoInductive` types,
such as streams, whose constructors can be infinitely nested. In Coq,
:cmd:`Variant` types thus correspond to the common subset of inductive
and coinductive types that are non-recursive.
```

jfehrle commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
 Variants
 --------
 
+The :cmd:`Variant` command allows defining types by listing
+the :term:`inhabitants <inhabitant>` of the type. This includes enumerated types from
+programming languages, such as Booleans, characters or even the
+degenerate cases of the unit and empty types. Variant types more
+generally include enumerated types with arguments or even enumerated
+types with parametric arguments such as option types, sum types,
+etc. It also includes predicates or type families defined by cases
+such as the Boolean reflection or equality predicates. Observing the
+form of the :term:`inhabitants <inhabitant>` of a variant type is done by case analysis
+using the :cmd:`match` expression. Each case is described by what is
+called a :gdef:`constructor` of the type. For instance, Booleans have two
+constructors: :g:`true` and :g:`false`.
```
```suggestion
The :cmd:`Variant` command allows defining types by listing
the :term:`inhabitants <inhabitant>` of the type.  This includes enumerated types from
programming languages, such as Booleans, characters or even the
degenerate cases of the unit and empty types. Variant types more
generally include enumerated types with arguments or even enumerated
types with parametric arguments such as option types and sum types.
It also includes predicates or type families defined by cases
such as the Boolean reflection or equality predicates. Observing the
form of the :term:`inhabitants <inhabitant>` of a variant type is done by case analysis
using the :cmd:`match` expression. Each case is described by what is
called a :term:`constructor` of the type. For instance, Booleans have two
constructors: :g:`true` and :g:`false`.
```

herbelin replied:
Adopting "etc" part.

For constructor, don't we need `gdef` to define it?

jfehrle replied:
Yes, `gdef`.  [Homer nods :-)](https://en.wiktionary.org/wiki/even_Homer_nods#English)

jfehrle commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
+etc. It also includes predicates or type families defined by cases
+such as the Boolean reflection or equality predicates. Observing the
+form of the :term:`inhabitants <inhabitant>` of a variant type is done by case analysis
+using the :cmd:`match` expression. Each case is described by what is
+called a :gdef:`constructor` of the type. For instance, Booleans have two
+constructors: :g:`true` and :g:`false`.
```
Constructors apply equally to Variants, Inductives and CoInductives, so I think they should be described at the beginning of inductive.rst.

Probably should say constructors have names.

herbelin replied:
Do you see a particular reason to define it at the beginning of inductive.rst rather than at the beginning of variants.rst (which comes first)?

jfehrle replied:
It will be fine in variants.rst.  (I looked at the order in which the changes appeared in github rather than in the actual manual.  I just got back to California Tuesday evening, so I'm a bit more addled than usual...)

jfehrle reviewed (commented):
Some earlier comments that got lost.

I think this still needs considerable work, but I don't want to spend the time right now to dig deeply into it.  So let's declare victory, merge it and move on.

jfehrle commented:
> Any objection to let `witness` out of the index?

It's worth mentioning once in the doc and keeping the index entry because the term is commonly used.

Zimmi48 commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
+etc. It also includes predicates or type families defined by cases
+such as the Boolean reflection or equality predicates. Observing the
+form of the :term:`inhabitants <inhabitant>` of a variant type is done by case analysis
+using the :cmd:`match` expression. Each case is described by what is
```
`match` is not a command. This is what makes the current build fail:

```suggestion
using the `match` expression. Each case is described by what is
```

herbelin commented:
"Witness": I'm unsure it is commonly used. I would say that "witness" generally refers to the connective `exists`, which is not part in itself of the logical foundation of Coq. (Some people use it to refer to a program realizing a proof but that in the context of the field of "realizability" which we do not refer to in the manual.)

herbelin commented:
> Probably worth mentioning the use of Inductive to create XX_ind and XX_rect as an alternative in the description of Variant.

That's a difficult question. From the recursive/non-recursive point of view, it is not justified to use suffix _ind or _rect for variants. The terminology "_case" would be more to the point.

Before the introduction of coinductive types, the term inductive type was a convenient way to characterize both variants and (properly-speaking) inductive types. So `_ind` was making sense at this time for variant types. But the introduction of coinductive types broke this view.

So, yes, something should be said. I tried the following:
```
. note::
  It is common in the standard library to use :cmd:`Inductive` in
  place of :cmd:`Variant` even for non-recursive types for the only
  purpose of automatically deriving associated schemes of name
  :n:`@ident`\ ``_rect``, :n:`@ident`\ ``_ind``, :n:`@ident`\
  ``_rec`` and :n:`@ident`\ ``_sind``.
```

jfehrle commented:
> "Witness": I'm unsure it is commonly used. I would say that "witness" generally refers to the connective exists, which is not part in itself of the logical foundation of Coq. (Some people use it to refer to a program realizing a proof but that in the context of the field of "realizability" which we do not refer to in the manual.)

Your choice.  I do hear Coq folk use the term, so it would be nice to mention.  I assumed it was synonymous with "inhabitant", but maybe that's not quite right (or not right at all).  If that's not right, maybe worth clarifying at some point.

jfehrle commented on doc/sphinx/language/core/inductive.rst:
```diff
@@ -1,6 +1,22 @@
 =======================================
 
+The :cmd:`Inductive` command allows defining types by cases on the form of the
+:term:`inhabitants <inhabitant>` of the type. In contrast, in types defined by the
+:cmd:`Variant` command, constructors can recursively have arguments in
+the type being defined. Inductive types include natural numbers,
```
Misworded?  Variants are not recursive (as stated a few sentences below).

herbelin replied:
I think, before, there was "On the contrary of types defined by the ..." but I did not realize that the changes that you proposed changed the meaning. Can I let you do a suggestion?

jfehrle replied:
How about this?

```suggestion
:term:`inhabitants <inhabitant>` of the type. These constructors can recursively
have arguments in the type being defined.  In contrast, in types defined by the
:cmd:`Variant` command, such recursive references are not permitted.
Inductive types include natural numbers,
```

herbelin replied:
Yes, I like it.

jfehrle commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -24,6 +43,51 @@ Variants
+  :term:`Leibniz Equality` is another example of variant type.
+
+.. note::
+   It is common in the standard library to use :cmd:`Inductive` in
+   place of :cmd:`Variant` even for non-recursive types for the only
+   purpose of automatically deriving associated schemes of name
+   :n:`@ident`\ ``_rect``, :n:`@ident`\ ``_ind``, :n:`@ident`\
+   ``_rec`` and :n:`@ident`\ ``_sind``.
```
```suggestion
   The standard library commonly uses :cmd:`Inductive` in
   place of :cmd:`Variant` even for non-recursive types in order to
   automatically derive the schemes
   :n:`@ident`\ ``_rect``, :n:`@ident`\ ``_ind``, :n:`@ident`\
   ``_rec`` and :n:`@ident`\ ``_sind``.  (These schemes are also created
   for :cmd:`Variant` if the :flag:`Nonrecursive Elimination Schemes` flag is set.)
```

jfehrle reviewed (commented):
Looks OK.  A couple further suggestions

jfehrle reviewed (approved):
Good enough for now.  @Zimmi48, any further comments?

Zimmi48 commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -24,6 +43,52 @@ Variants
+
+     End FreshNameSpace.
+
+  :term:`Leibniz Equality` is another example of variant type.
```
Casing matters (this is making the build fail):

```suggestion
  :term:`Leibniz equality` is another example of variant type.
```

herbelin commented:
If no other comments, I'll squash.

Zimmi48 commented on doc/sphinx/language/core/inductive.rst:
```diff
@@ -1,6 +1,23 @@
+
+Due to the recursive structure of inductive types, functions on
+inhabitants of an inductive type have in general to be defined
+recursively. This is provided by the :n:`fix` expression and the
```
```suggestion
recursively. Support for defining recursive functions is provided by the :n:`fix` expression and the
```

or:

```suggestion
recursively. This is done using the :n:`fix` expression and the
```

herbelin replied:
Using the proposal of @jfehrle instead.

Zimmi48 reviewed (approved):
LGTM, thanks!

At some point, we should move the explanation on parameters and constructors currently found under `Inductive` to the description of `Variant`, but let's keep this as future work.

jfehrle commented:
> At some point, we should move the explanation on parameters and constructors currently found under `Inductive` to the description of `Variant`, but let's keep this as future work.

We should consider grouping the sections Variants, Record Types, Inductive Types and Coinductive Types into a new section "Defining New Types" in a future PR.  They go together and there are too many sections under Core Language.  WDYT?

jfehrle commented:
Not sure what to make of this.  Does it just mean that you must use coqbot?

![image](https://github.com/coq/coq/assets/1253341/bc383445-904c-47dc-9666-ca008544f0a6)

proux01 commented:
> Not sure what to make of this. Does it just mean that you must use coqbot?

Exactly, this is just to prevent accidentally pushing the merge button.

proux01 commented:
> > At some point, we should move the explanation on parameters and constructors currently found under `Inductive` to the description of `Variant`, but let's keep this as future work.
> 
> We should consider grouping the sections Variants, Record Types, Inductive Types and Coinductive Types into a new section "Defining New Types" in a future PR. They go together and there are too many sections under Core Language. WDYT?

No strong opinion but if done we should find a more precise title, many types are neither variant nor {co,}inductive.

jfehrle commented on doc/sphinx/language/core/inductive.rst:
```diff
@@ -1,6 +1,23 @@
+:cmd:`Variant` types thus correspond to the common subset of inductive
+and coinductive types that are non-recursive.
+
+Due to the recursive structure of inductive types, functions on
+inhabitants of an inductive type have in general to be defined
+recursively. This is provided by the :n:`fix` expression and the
```
```suggestion
Due to the recursive structure of inductive types, functions on
inductive types generally must be defined
recursively using the :n:`fix` expression (see :n:`@term_fix`) or the
```

WDYT?

jfehrle commented:
> No strong opinion but if done we should find a more precise title, many types are neither variant nor {co,}inductive.

What are some of the other ways to create new types that I'm forgetting?  Even so, I think grouping all the ways you can explicitly create new types makes sense.  What do you find imprecise about "Defining New Types"?

proux01 commented:
For instance `Definition nat_fun := nat -> nat.` defines a new type.

Zimmi48 commented:
> > At some point, we should move the explanation on parameters and constructors currently found under `Inductive` to the description of `Variant`, but let's keep this as future work.
> 
> We should consider grouping the sections Variants, Record Types, Inductive Types and Coinductive Types into a new section "Defining New Types" in a future PR. They go together and there are too many sections under Core Language. WDYT?

Well, I disagree. One of the benefits of the restructuring was to separate different notions into separate pages to get smaller pages. It would feel backwards to me to merge them again.

jfehrle commented:
I'm suggesting a small change to the table of contents, not to combine the existing pages.  Just adding a new intermediate node in the TOC tree and a maybe a very brief introduction.

Zimmi48 commented:
> I'm suggesting a small change to the table of contents, not to combine the existing pages. Just adding a new intermediate node in the TOC tree and a maybe a very brief introduction.

Ah, is that even supported by Sphinx? If yes, this would create a small issue which would be that not all the refman would have its leaf pages at the same level.

Zimmi48 commented:
@herbelin Wasn't this almost ready to be squashed and merged?

herbelin commented:
I fixed a few more things (including a sentence finally telling the opposite of what was intended) and squashed. If you can quickly double-check, that'll be nice.

herbelin commented:
Sorry, forget my last comment, I forgot to pull (the misinterpretation was fixed). I'm going to re-push.

jfehrle commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
 Variants
 --------
 
+The :cmd:`Variant` command allows defining types by listing
+the :term:`inhabitants <inhabitant>` of the type. This includes enumerated types from
+programming languages, such as Booleans, characters or even the
```
```suggestion
The :cmd:`Variant` command allows defining types by listing
the :term:`inhabitants <inhabitant>` of the type.  Each inhabitant is
specified by a :gdef:`constructor`.  For instance, Booleans have two
constructors: :g:`true` and :g:`false`. Types can include enumerated types from
programming languages, such as Booleans, characters or even the
```

Moved two sentences up from below and simplified wording for constructors.  (not sure why the diff shown is not more granular today)

herbelin replied:
Yes, much better.

jfehrle commented on doc/sphinx/language/core/variants.rst:
```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
+It also includes predicates or type families defined by cases
+such as the Boolean reflection or equality predicates. Observing the
+form of the :term:`inhabitants <inhabitant>` of a variant type is done by case analysis
+using the `match` expression. Each case is described by what is
+called a :gdef:`constructor` of the type. For instance, Booleans have two
+constructors: :g:`true` and :g:`false`.
```
```suggestion
using the `match` expression.
```

jfehrle commented on doc/sphinx/language/extensions/evars.rst:
```diff
@@ -189,7 +189,7 @@ automatically as a side effect of other tactics.
       exact H1. (* resolves the first goal and by side effect ?x and ?m *)
 
    The :n:`?x` and :n:`?m` goals ask for proof that :n:`nat` has a
```
```suggestion
   The :n:`?x` and :n:`?m` goals ask for proof that :n:`nat` has an
```

herbelin replied:
Well spotted.

jfehrle reviewed (approved):
LGTM.  Final wording suggestions below.

jfehrle commented:
@herbelin Unless someone has further comments, how about squashing, then someone can merge?

herbelin commented:
It was squashed but I forgot to remove the label.

jfehrle commented:
@coqbot: merge now
